egs/wsj0-mix-var/Multi-Decoder-DPRNN/train.py

```diff
@@ -0,0 +1,214 @@
+    checkpoint = ModelCheckpoint(
+        dirpath=checkpoint_dir,
+        filename="{epoch}-{step}",
+        monitor="avg_sdr",
+        mode="max",
+        save_top_k=5,
+        verbose=True,
+    )
+    callbacks.append(checkpoint)
+    if conf["training"]["early_stop"]:
+        callbacks.append(EarlyStopping(monitor="avg_sdr", mode="max", patience=30, verbose=True))
+
+    # Don't ask GPU if they are not available.
+    gpus = -1 if torch.cuda.is_available() else None
+    distributed_backend = "dp" if torch.cuda.is_available() else None
+
+    # Train model
+    trainer = pl.Trainer(
+        max_epochs=conf["training"]["epochs"],
+        callbacks=callbacks,
+        default_root_dir=exp_dir,
+        gpus=gpus,
+        distributed_backend=distributed_backend,
+        limit_train_batches=1.0,  # Useful for fast experiment
+        gradient_clip_val=200,
+        resume_from_checkpoint=conf["main_args"]["resume_from"],
+    )
+    trainer.fit(system)
+
+    best_k = {k: v.item() for k, v in checkpoint.best_k_models.items()}
+    with open(os.path.join(exp_dir, "best_k_models.json"), "w") as f:
+        json.dump(best_k, f, indent=0)
+    # Save last model for convenience
+    torch.save(system.model.state_dict(), os.path.join(exp_dir, "final_model.pth"))
+
+
+class VarSpkrSystem(System):
+    def common_step(self, batch, batch_nb, train=True):
+        mixture_tensor, source_tensor, ilens, num_sources = batch
+        pred_tensor, selector_output = self(mixture_tensor, num_sources.tolist())
+        batch_size, num_stages, _, T = pred_tensor.size()
+        avg_loss = 0
+        spks_sdr = []
+        accuracy = 0
+        for i in range(batch_size):
+            est_src = pred_tensor[i, :, : num_sources[i], : ilens[i]]
+            src = source_tensor[i, : num_sources[i], : ilens[i]]
+            logits = selector_output[i]
+            loss, pos_sdr, correctness = self.loss_func(est_src, logits, src)
+            avg_loss = avg_loss + loss / batch_size
+            spks_sdr.append((num_sources[i], pos_sdr))
+            accuracy += correctness / batch_size
+        return avg_loss, spks_sdr, accuracy
+
+    def training_step(self, batch, batch_nb):
+        avg_loss, spks_sdr, accuracy = self.common_step(batch, batch_nb, train=True)
+        self.log("loss", avg_loss)
+        for num_spks, sdr in spks_sdr:
+            self.log(f"{num_spks}spks_sdr_tr", sdr)
+        self.log("acc_tr", accuracy)
+        return avg_loss
+
+    def validation_step(self, batch, batch_nb):
+        avg_loss, spks_sdr, accuracy = self.common_step(batch, batch_nb, train=False)
+        self.log("val_loss", avg_loss)
+        for num_spks, sdr in spks_sdr:
+            self.log(f"{num_spks}spks_sdr_val", sdr)
+        self.log("acc_val", accuracy)
+        # SDR averaged across number of sources
+        avg_sdr = torch.mean(torch.Tensor([sdr for _, sdr in spks_sdr]))
+        self.log("avg_sdr", avg_sdr)
+
+
+class WeightedPITLoss(nn.Module):
```

Comment:
Can you add a docstring and some comments for understanding what this does on the high level?

Reply:
Is this fixed?

Reply:
Yeah, you can see the docstring above the __init__ line.